src/routers/SwitchRouter.js

```diff
@@ -88,35 +88,56 @@ export default (routeConfigs, config = {}) => {
     };
   }
 
-  function getNextState(prevState, possibleNextState) {
-    if (!prevState) {
-      return possibleNextState;
+  function getNextState(action, prevState, possibleNextState) {
+    function updateNextStateHistory(nextState) {
+      if (backBehavior !== 'history') {
+        return nextState;
+      }
+      let nextRouteKeyHistory = prevState.routeKeyHistory;
+      if (action.type === NavigationActions.NAVIGATE) {
+        nextRouteKeyHistory = [...prevState.routeKeyHistory]; // copy
+        const keyToAdd = nextState.routes[nextState.index].key;
+        nextRouteKeyHistory = nextRouteKeyHistory.filter(k => k !== keyToAdd); // dedup
+        nextRouteKeyHistory.push(keyToAdd);
+      } else if (action.type === NavigationActions.BACK) {
+        nextRouteKeyHistory = [...prevState.routeKeyHistory]; // copy
+        nextRouteKeyHistory.pop();
+      }
+      return {
+        ...nextState,
+        routeKeyHistory: nextRouteKeyHistory,
+      };
     }
 
-    let nextState;
-    if (prevState.index !== possibleNextState.index && resetOnBlur) {
+    let nextState = possibleNextState;
+    if (
+      prevState &&
+      prevState.index !== possibleNextState.index &&
+      resetOnBlur
+    ) {
       const prevRouteName = prevState.routes[prevState.index].routeName;
       const nextRoutes = [...possibleNextState.routes];
       nextRoutes[prevState.index] = resetChildRoute(prevRouteName);
-
-      return {
+      nextState = {
         ...possibleNextState,
         routes: nextRoutes,
       };
-    } else {
-      nextState = possibleNextState;
     }
-
-    return nextState;
+    return updateNextStateHistory(nextState);
```

Comment:
I simplified a bit the getNextState original code, it's clearer to have a single return path imho